### PR TITLE
[SPARK-47496][SQL] Java SPI Support for dynamic JDBC dialect registering

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

This PR brings the Java ServiceProvider Interface (SPI) Support for dynamic JDBC dialect registering.

A custom JDBC dialect can be registered easily instead of calling JdbcDialects.registerDialect manually.

### Why are the changes needed?

For pure SQL and other non-Java API users, it's difficult to register a custom JDBC dialect to use. With this patch, this can be done when the jar containing the dialect class is visible to the spark classloader.

### Does this PR introduce _any_ user-facing change?

Yes, but mostly for third-party developers

### How was this patch tested?

new tests

### Was this patch authored or co-authored using generative AI tooling?
no

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/OracleDialect.scala [hunk 1]
```diff
@@ -230,3 +224,11 @@ private case object OracleDialect extends JdbcDialect {
 
   override def supportsOffset: Boolean = true
 }
+
+private[jdbc] object OracleDialect {
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/OracleDialect.scala:228]: Ur, actually, `case object` has more features (serializable, hashCode, toString) than `object`. Do we need to shrink features like this?

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala [hunk 2]
```diff
@@ -2154,4 +2155,13 @@ class JDBCSuite extends QueryTest with SharedSparkSession {
     val expected = Map("percentile_approx_val" -> 49)
     assert(namedObservation.get === expected)
   }
+
+
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala:2159]: extra empty line?

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM (with minor comments). It's a nice feature, @yaooqinn !

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/OracleDialect.scala (hunk 1 above)
Comment (yaooqinn) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/OracleDialect.scala:228]: This is only a place for containing some const ints, which represent some variants from Oracle JDBC extentions. Thus, `object` is enough

**Comment by yaooqinn** [comment]
Thank you @dongjoon-hyun.

I addressed the comments

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/OracleDialect.scala (hunk 1 above)
Comment (yaooqinn) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/OracleDialect.scala:228]: Opps, the errors occured
```
- simple scan with LIMIT *** FAILED *** (65 milliseconds)
[info]   org.apache.spark.SparkException: Task not serializable
```

**Comment by dongjoon-hyun** [comment]
https://github.com/apache/spark/pull/45626#discussion_r1533280693 seems to be missed still. Are you going to recover it, @yaooqinn ?
> I addressed the comments

**Comment by yaooqinn** [comment]
Yes @dongjoon-hyun it was caused by the unserializable object `SpecificTypes` inside MsSqlServerDialect

**Comment by beliefer** [comment]
@yaooqinn It's really worth? as I know, users could register dialects with JdbcDialects.registerDialect in V2 catalog.
Users can implement CatalogPlugin and register the catalog and the dialects.

**Comment by yaooqinn** [comment]
@beliefer Yes, it's worth it.

Actually, there are plenty of hacky ways to register dialects just like what you mentioned. We can't anticipate those as typical API usages.

**Comment by yaooqinn** [comment]
Merged to master.

Thank you @dongjoon-hyun

**Comment by smileyboy2019** [comment]
In which version will it be released, and can multiple libraries be joined for querying?

**Comment by elijah-pl** [comment]
@yaooqinn is it only in Spark 4 or 3 as well?

**Comment by yaooqinn** [comment]
@elijah-pl please refer to the JIRA sizde for fixed versions

**Pull request closed without merging**